aggregator/src/aggregator.rs

```diff
@@ -1179,19 +1226,52 @@ impl VdafOps {
                 }
             };
 
-        let encoded_leader_plaintext_input_share = match hpke::open(
-            hpke_keypair.config(),
-            hpke_keypair.private_key(),
-            &HpkeApplicationInfo::new(&Label::InputShare, &Role::Client, task.role()),
-            leader_encrypted_input_share,
-            &InputShareAad::new(
-                *task.id(),
-                report.metadata().clone(),
-                report.public_share().to_vec(),
-            )
-            .get_encoded(),
-        ) {
-            Ok(encoded_leader_plaintext_input_share) => encoded_leader_plaintext_input_share,
+        let try_hpke_open = |hpke_keypair: &HpkeKeypair| -> Result<Vec<u8>, Error> {
+            match hpke::open(
+                hpke_keypair.config(),
+                hpke_keypair.private_key(),
+                &HpkeApplicationInfo::new(&Label::InputShare, &Role::Client, task.role()),
+                leader_encrypted_input_share,
+                &InputShareAad::new(
+                    *task.id(),
+                    report.metadata().clone(),
+                    report.public_share().to_vec(),
+                )
+                .get_encoded(),
+            ) {
+                Ok(results) => Ok(results),
+                Err(error) => Err(error.into()),
+            }
+        };
+
+        let global_hpke_keypair =
+            global_hpke_keypairs.keypair(leader_encrypted_input_share.config_id());
+
+        let task_hpke_keypair = task
+            .hpke_keys()
+            .get(leader_encrypted_input_share.config_id());
+
+        let decryption_result = match (task_hpke_keypair, global_hpke_keypair) {
+            // Verify that the report's HPKE config ID is known.
+            // https://www.ietf.org/archive/id/draft-ietf-ppm-dap-02.html#section-4.3.2
+            (None, None) => {
+                return Err(Arc::new(Error::OutdatedHpkeConfig(
+                    *task.id(),
+                    *leader_encrypted_input_share.config_id(),
+                )));
+            }
+            (None, Some(global_hpke_keypair)) => try_hpke_open(&global_hpke_keypair),
+            (Some(task_hpke_keypair), None) => try_hpke_open(task_hpke_keypair),
+            (Some(task_hpke_keypair), Some(global_hpke_keypair)) => {
+                match try_hpke_open(task_hpke_keypair) {
+                    Ok(plaintext) => Ok(plaintext),
+                    Err(_) => try_hpke_open(&global_hpke_keypair),
```

Comment:
I think we should only fall back to the global HPKE keypair if the error from `try_hpke_open` is `hpke::Error::Hpke`, because otherwise we might suppress evidence of some other problem.